### PR TITLE
vSphere Cloud Provider: add SAML token authentication support

### LINKED ISSUE #63209: Support SAML token authentication in vSphere Cloud Provider
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

/sig vmware

**What happened**:

The vSphere cloud provider currently supports username+password based authentication, where the credentials must be present in plain text within the cloud
provider configuration file.

**What you expected to happen**:

Authenticate with vSphere SDK using a SAML token instead.

### PR DESCRIPTION
**Pull request by dougm**
**What this PR does / why we need it**:

The vSphere cloud provider currently supports username+password based authentication, this PR adds an option to use token based authentication.

**Which issue(s) this PR fixes**:

Fixes #63209

**Special notes for your reviewer**:

For now the config structs and validation are left as-is and
the LoginByToken method is used if the username value is PEM encoded.
In this case of username field configured with the public key, the password
field is expected to be configured with the private key.

In a follow-up PR we can look at collapsing the auth related fields into
a common struct to avoid duplication of field merging and validation.
And then add separate fields for the public and private keys.

**Release note**:

```release-note
vSphere Cloud Provider: add SAML token authentication support
```

### CODE REVIEW

File: pkg/cloudprovider/providers/vsphere/vclib/connection.go [hunk 1]
```diff
@@ -78,6 +81,49 @@ func (connection *VSphereConnection) Connect(ctx context.Context) error {
 	return nil
 }
 
+// login calls SessionManager.LoginByToken if certificate and private key are configured,
+// otherwise calls SessionManager.Login with user and password.
+func (connection *VSphereConnection) login(ctx context.Context, client *vim25.Client) error {
+	m := session.NewManager(client)
+
+	// TODO: Add separate fields for certificate and private-key.
```
Comment (abrarshivani) [inline pkg/cloudprovider/providers/vsphere/vclib/connection.go]: Is this issue tracked somewhere?

**Review by abrarshivani — APPROVED** [review]
/lgtm

File: pkg/cloudprovider/providers/vsphere/vclib/connection.go (hunk 1 above)
Comment (dougm) [reply pkg/cloudprovider/providers/vsphere/vclib/connection.go]: Not yet, I'll create a new issue.

**Comment by dougm** [comment]
/assign @cblecker @thockin @sttts 

@ dep-approvers can you review the Godeps updates?

**Comment by cblecker** [comment]
There's some new imports, but they're all under the same repo/dep, and all Apache 2.0 licensed.

/approve

**Comment by smarterclayton** [comment]
This looks like it has data races:

```
=== RUN   TestVSphereLoginByToken
E0705 17:13:06.921738    9479 connection.go:134] Logout failed: Post https://127.0.0.1:36902/sdk: dial tcp 127.0.0.1:36902: connect: connection refused
==================
WARNING: DATA RACE
Write at 0x00c4200fbbc0 by goroutine 48:
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vclib.(*VSphereConnection).Connect()
      /go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vclib/connection.go:59 +0x3ff
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere.TestVSphereLoginByToken()
      /go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vsphere_test.go:246 +0x335
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Previous read at 0x00c4200fbbc0 by goroutine 5:
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere.logout()
      /go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vsphere.go:496 +0x13b

Goroutine 48 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:978 +0x2cd
  main.main()
      _testmain.go:112 +0x324

Goroutine 5 (running) created at:
  runtime.createfing()
      /usr/local/go/src/runtime/mfinal.go:156 +0x61
  os.NewFile()
      /usr/local/go/src/os/file_unix.go:79 +0x55
  os.init()
      /usr/local/go/src/os/file.go:58 +0x418
  main.init()
      <autogenerated>:1 +0x96
==================
--- FAIL: TestVSphereLoginByToken (0.31s)
	testing.go:730: race detected during execution of test
```

**Comment by dougm** [comment]
@smarterclayton PR with fix #65913

**Pull request merged**